### ISSUE TITLE
Always link to the local libcaffe.so under ./build/lib

### ISSUE THREAD

**Issue by orsharir** [BVLC/caffe#3385]
I was recently faced with a weird compilation error on my university's system, while trying to compile caffe on my home directory. The source of the error was that during the linking phase all files under the "tools" directory tried to link against an old caffe library installed system-wide in /usr/local/lib. To solve this issue I had to manually add ./build/lib to my Makefile.config's LIBRARY_DIRS variable, and make sure it was placed before /usr/local/lib.

While the above scenario might be very specific to my university, it still seems to be more correct to link against the local library of caffe under build/lib by default, either by changing Makefile.config.example or by changing the Makefile itself. I think the latter is the preferred option.

**Comment by xiangqiaolxq** [comment]
I have meet the same problem because I set the LD_LIBRARY_PATH which indicate to link old caffe library .Check your own system LD_LIBRARY_PATH,delete related may help...

**Comment by orsharir** [comment]
The problem is that if caffe's library is installed in /usr/local/lib, and let's say you also work on a development version of caffe and you try to build any of the executables, then by default Caffe will link against the library in /usr/local/lib instead the local one in ./build/lib.

**Comment by xiangqiaolxq** [comment]
Yeah, if libcaffe.so exist in /usr/local/lib, this problem may happen. When create libcaffe.so, system will search /usr/local/lib before the directory ./build/lib. Thus caffe's library (libcaffe.so) link wrong .
I set this option LD_LIBRARY_PATH=${CAFFE_HOME}/.build/lib:$LD_LIBRARY_PATH ,all things down.
Maybe you can also try to rename /usr/local/libcaffe.so to another name like /usr/local/libcaffe_bak.so then build your caffe. After that, rename /usr/local/libcaffe_bak.so back to /usr/local/libcaffe_bak.so .
